### PR TITLE
Add FrameState and add pending pipelines to Frame immediately

### LINKED ISSUE #12784: joint_session should include pending pipelines
If the user presses the back button too early, before a pending pipeline is added, we end up thawing/freezing the wrong pipelines.

Run with: `./mach run -d -- -z --pref dom.mozbrowser.enabled /tmp/a.html`

``` html
<iframe mozbrowser="true"></iframe>

<script>
  var iframe = document.querySelector("iframe");
  var once = (elt, event, cb) => {
    var _cb = () => {
      elt.removeEventListener(event, _cb);
      cb();
    }
    elt.addEventListener(event, _cb);
  }
  iframe.addEventListener("mozbrowserloadstart", e => console.log("START"));
  iframe.addEventListener("mozbrowserloadend", e => console.log("END"));
  once(iframe, "mozbrowserloadend", () => {

    var back = () => iframe.goBack();

    once(iframe, "mozbrowserloadstart", back);

    iframe.src = "http://example.com/test";
  });
  iframe.src = "http://example.com";
</script>
```

```
ERROR:servo: assertion failed: self.suspended_since.get().is_none() (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at /Users/Paul/git/servo/components/script/timers.rs:230)
ERROR:servo: called `Option::unwrap()` on a `None` value (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }, at ../src/libcore/option.rs:326)
```

In the above test case, we have 4 pipelines:
- 0,0 top level document, we don't care
- 1,0 about:blank
- 1,1 http://example.com
- 1,2 http://example.com/test

On `goBack()`, `traverse_frame_to_pipeline` is called with 1,0 instead of 1,1.

/cc @ConnorGBrewster

### PR DESCRIPTION
**Pull request by cbrewster**
r? @asajeffrey 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #12784 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because refactoring

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12810)

### CODE REVIEW

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -2083,9 +2059,12 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
             return ReadyToSave::NoRootFrame;
         }
 
-        // If there are pending loads, wait for those to complete.
-        if !self.pending_frames.is_empty() {
-            return ReadyToSave::PendingFrames;
+        if let Some(root_frame_id) = self.root_frame_id {
+            for frame in self.full_frame_tree_iter(root_frame_id) {
```
Comment (ConnorGBrewster) [inline components/constellation/constellation.rs]: Not sure of the performance consequences here when doing reftests. There shouldn't be many pipelines/frames.

**Comment by ConnorGBrewster** [comment]
Lets see if this adds any breakage.
@bors-servo try

**Comment by ConnorGBrewster** [comment]
Another issue to think about is when we activate the new document, we traverse to it. This means that any traversals that happen while the document is immature will be overwritten by the traversal when the document is activated.

**Comment by asajeffrey** [comment]
Yes, we'd need to replace the on-load handler by one which updates the pipeline state rather than traversing to it.

**Comment by ConnorGBrewster** [comment]
An alternate approach would be to add pending pipelines to `frame.current`. And when pushing a new pending pipelines, check current to see if it is pending, if so abort that load and replace it with the new pending pipeline.

**Comment by ConnorGBrewster** [comment]
@bors-servo try
@asajeffrey this now adds pending pipelines to `frame.current` instead of pushing them on `frame.next` and traversing when they mature.

File: components/constellation/constellation.rs [hunk 2]
```diff
@@ -219,42 +216,63 @@ pub struct InitialConstellationState {
...
             next: vec!(),
         }
     }
 
-    fn load(&mut self, pipeline_id: PipelineId) {
-        self.prev.push(self.current);
-        self.current = (pipeline_id, Instant::now());
+    fn load(&mut self, pipeline_id: PipelineId) -> Option<FrameState> {
+        let current = replace(&mut self.current, FrameState::new(pipeline_id, self.id));
+        // TODO(ConnorGBrewster): In my opinion, if the current state is pending, we should
```
Comment (ConnorGBrewster) [inline components/constellation/constellation.rs]: @asajeffrey what is your opinion on this?

**Comment by ConnorGBrewster** [comment]
Not quite sure why the root browsing context lookup here is failing. I cannot reproduce these failures on my machine. 😞 I will look more in depth on Monday

**Comment by asajeffrey** [comment]
Reviewed 1 of 1 files at r1.
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/constellation/constellation.rs, line 2063 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12810#-KOqiFbTx_NNxxgeqtMW-r2-2063:-KPDeOxZskk2VgfhfY6s:b-an1rjg) ([raw file](https://github.com/servo/servo/blob/fedbc6df1e895a2c40f42ca437e583ffb55502fa/components/constellation/constellation.rs#L2063)):*
<details><summary><i>Previously, ConnorGBrewster (Connor Brewster) wrote…</i></summary>
> Not sure of the performance consequences here when doing reftests. There shouldn't be many pipelines/frames.
</details>
Indeed. The frame tree is usually pretty shallow. I guess the history could get long after a long browsing session, but still I wouldn't expect this to be on the hot path.

---

*[components/constellation/constellation.rs, line 260 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12810#-KOqiFbTx_NNxxgeqtMW-r3-260:-KPDdohkh3Jc4Re8O0m_:bm4t7rv) ([raw file](https://github.com/servo/servo/blob/102732072fb26b1d5212cad55a5956aec7043cab/components/constellation/constellation.rs#L260)):*
<details><summary><i>Previously, ConnorGBrewster (Connor Brewster) wrote…</i></summary>
> @asajeffrey what is your opinion on this?
</details>
I think I'm a lot happier with a model that we set the current frame immediately, much less opportunity for race conditions. I'm a bit concerned that there may be other code that expects frames only to be set when they are loaded.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12810)*

**Comment by asajeffrey** [comment]
The changes look good, we just need to work out if there's any code that assumes frames are added to the frame tree only after loading.

**Comment by ConnorGBrewster** [comment]
@asajeffrey sounds good, the panics found from the testing may be due to something relying on the old behavior.

**Comment by asajeffrey** [comment]
The failure of root browsing context lookup is presumably a timing issue, that it assumes the lookup doesn't happen until after the page is loaded.

**Comment by ConnorGBrewster** [comment]
True, the constellation may be sending its script thread a message that causes the lookup. This is probably due to the constellation assuming the current pipeline in a frame is not pending

**Comment by nox** [comment]
This looks legit.

**Comment by ConnorGBrewster** [comment]
Closing until I can figure out a way to prevent the lookup failure.

**Pull request closed without merging**